brave/brave-core
# Do not overwrite Rewards wallet data on read error

Resolves https://github.com/brave/brave-browser/issues/16373

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:

### Scenario: App loses access to system password store

- Given that the user has opted in to Rewards and created a rewards wallet
- When the app loses access to the operating system's "keychain"
- Then the existing wallet data is not overwritten
- And an error indicating decryption failure is written to the Rewards error log.

### Scenario: Rewards wallet data is corrupted

- Given that the user has opted in to Rewards and created a rewards wallet
- When the `brave.rewards.wallets.brave` preferences is corrupted
- Then the existing wallet data is not overwritten
- And an error indicating the read failure is written to the Rewards error log

#### Possible test steps:

- Open the browser with a clean profile on staging.
- Open the rewards panel and click "Start using Brave Rewards" to generate a rewards wallet.
- Accept the UGP grant.
- Close the browser.
- Remove the browser's access to the system password store.
  - On MacOS, this can be accomplished by opening the "Keychain Access" app and removing the app from the access list for "Brave Safe Storage".
- Start the browser and deny the app's request to access the system password store.
- Navigate to `brave://rewards`
- Attempt to toggle the Ads enabled setting (previously, this would generate a new rewards wallet).
- Navigate to `brave://rewards-internals` and view the logs.
  - Verify that a "Decryption failed" message has been logged.
- Restart the browser, and grant access to the system password store.
  - Verify that Rewards functionality works (i.e. that the user can tip).
